src/include/migraphx/shape_transform_descriptor.hpp

```diff
@@ -0,0 +1,112 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#ifndef MIGRAPHX_GUARD_MIGRAPHX_SHAPE_TRANSFORM_DESCRIPTOR_HPP
+#define MIGRAPHX_GUARD_MIGRAPHX_SHAPE_TRANSFORM_DESCRIPTOR_HPP
+
+#include <migraphx/config.hpp>
+#include <migraphx/optional.hpp>
+#include <cstdint>
+#include <vector>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct operation;
+
+// The shape_transform_descriptor class is data structure to simplify shape
+// transformations like reshape, transpose, broadcast, etc. This is made up
+// of a collection of dimensions which are a collection of subdimensions.
+//
+// Each subdimension has an axis and a `len`. The `len` is the length of the
+// subdimension. The axis represents the axis the dimension originated. It is
+// represented as a vector, the first element represents the axis in the
+// original dimension and the additional elements are used when such
+// dimension is split. The axis is empty when its a broadcasted dimension,
+// and a hidden axis can be set if the dimension is associated with a `1`
+// dimension in the original shape.
+//
+// This will first record shape transformations with the `apply` method. This
+// will manipulate this data structure to represent how the transformation
+// changes the dimensions.
+//
+// For example, if we start with an initial dimensions as `[x, y, z]` then
+// each dimension will have one subdimension that corresponds to each
+// original dimension: `[[x:0]], [[y:1]], [[z:2]]`.
+//
+// When a transpose is applied we would just permutate the dimensions.
+//
+// When a reshape that would merge dimensions together then the subdimensions
+// are copied to the same subdimension. So if we reshape the dimensions as `
+// [x*y, z]` then it would become `[[x:0], [y:1]], [[z:1]]`. If the reshape
+// splits the dimension then the subdimension is copied to each dimension and
+// the axis is updated to maintain the order. So a reshape of `[2, x/2, y,
+// z]` would become: `[[2:0,0]], [[x/2:0,1]], [[y:1]], [[z:2]]`.
```

Comment:
In `[[2:0, 0]]` what does second element mean `,0]]` ? 
`2:0` means subdimension with len 2 originating from axis 0.  but then what is second element ?

Reply:
> The first elements represents the original axis, 

Original axis of the original input before applying any transformation ? 
e.g. 
let's say I have input  of shape `[2, 5, 10]` --> which would be represented as  `[[2:0]], [[5:1]], [[10:2]]` in current scheme. 
Now, Applying first reshape it becomes `[2, 5, 5, 2]` --> which would become `[[2:0]], [[5:1]], [[5:2, 2]], [[2:2, 2]]` 
Now let's apply second reshape and make it `[2, 25, 2]` --> 

1. Original  axis here should it taken from original input `[2, 5, 10]` ? or 
In this case, output dimension with len 25 has two parent axes. 
2. Should it be considered from Input to the second reshape which is `[2, 5, 5, 2]`?

Or to frame question differently, 
what is the definition of original axis ?

Reply:
> but each additional element represents where the axis is split. 

To me it seems like size of `subdimension.axis` can at max be 2 and min 1. 

for the case when size is two. Second element would denote axis in the output of the operation. 

For example
> So a reshape of `[2, x/2, y,
// z]` would become: `[[2:0,0]], [[x/2:0,1]], [[y:1]], [[z:2]]`. 

here `[[x/2:0, 1]]` mean 1 is axis in the output for that dimension.   We can instead make the axis always of length 2. where first element would be parent axis and second element would be axis in the output. 
`[[2:0,0]], [[x/2:0,1]], [[y:1, 2]], [[z:2, 3]]`

> So if we reshape the dimensions as `
// [x*y, z]` then it would become `[[x:0], [y:1]], [[z:2]]`.

E.g. in this case it would be like `[[x:0, 0], [y:1, 0]], [[z:2, 1]]`

Reply:
> Now, Applying first reshape it becomes [2, 5, 5, 2] --> which would become [[2:0]], [[5:1]], [[5:2, 2]], [[2:2, 2]]

No, it becomes `[[2:0]], [[5:1]], [[5:2, 0]], [[2:2, 1]]`.

> Now let's apply second reshape and make it [2, 25, 2] 

So this would become `[[2:0]], [[5:1], [5:2, 0]], [[2:2, 1]]`.

> Original axis here should it taken from original input [2, 5, 10] ?

Yes in fact all the axes are from the original input we just add extra numbers to track the splitting of the axis.

> To me it seems like size of subdimension.axis can at max be 2 and min 1.

No that isnt the case. It can also be empty for when its unsqueezed and broadcasted.

It is a max of 2 only after `simplify` because it renumbers all axes, but before that it can be more than 2.

For example, if we have `[[12:0]]` that we reshape to `[2, 6]` so it would become `[[2:0, 0], [6:0, 1]]`. And then if there is another reshape to `[2, 2, 3]` then this will become `[[2:0, 0], [2:0, 1, 0], [3:0, 1, 1]]`.  

> here [[x/2:0, 1]] mean 1 is axis in the output for that dimension.

No, its just used to track the order of the dimensions after the axis is split by a reshape. You can see the logic here:

https://github.com/ROCm/AMDMIGraphX/blob/9a43b2bed0117080cfcf6ca29603417d7a976da4/src/shape_transform_descriptor.cpp#L198

We `push_back` the numbers from the range (0, n], where n is the number of splits that reshape is doing.